Redo tempo-cli with basic command structure and improvements

mdisibio opened the pull request:
**What this PR does**:
Redo of tempo-cli to have basic command structure for easier maintenance and new commands.  Existing three commands now look like:
- `tempo-cli list block <tenant-id> <block-id>`
- `tempo-cli list blocks <tenant-id>`
- `tempo-cli query <trace-id>`

Notable changes:
- Using kong library, provides help (-h, --help) and discoverability at each level (./tempo-cli -h, list -h, list blocks -h)
- Added universal --config-file (-c) parameter to load backend settings from existing tempo yaml.  --backend, etc existing parameters can still be used and override config values if both are used.
- Updated behavior of list block/blocks commands to put expensive operations behind additional flags:  By default only meta is loaded.  list blocks --load-index to check index for dupes and record counts, list block --check-dupes to load block and check for duplicate trace ids.  

**Which issue(s) this PR fixes**:
Fixes #281

**Checklist**
- [ ] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

joe-elliott commented:
Welcome improvement in both code organization and tempo-cli ux.  

The current main readme.md now contains some out of date info on tempo-cli.  What do you think about adding updated info to the `./docs` folder and just linking to it from the main readme?

mdisibio commented:
> Welcome improvement in both code organization and tempo-cli ux.
> 
> The current main readme.md now contains some out of date info on tempo-cli. What do you think about adding updated info to the `./docs` folder and just linking to it from the main readme?

Agree.  Added CLI section to docs that covers commands, options, etc, and updated readme to link to it.

annanay25 reviewed (approved):
Fantastic work, thank you @mdisibio! Can you please also add a CHANGELOG entry and then we can merge this!

achatterjee-grafana commented on docs/tempo/website/cli/_index.md:
````diff
@@ -0,0 +1,114 @@
+```bash
+tempo-cli command [subcommand] [options] [arguments...]
+```
+
````
```suggestion
Remove extra space.
```

achatterjee-grafana commented on docs/tempo/website/cli/_index.md:
```diff
@@ -0,0 +1,114 @@
+
+# Tempo CLI
+
+Tempo CLI is a separate executable that contains utility functions related to tempo.  It is not required for a working install, but may be helpful for analysis or troubleshooting.
```
```suggestion
Tempo CLI is a separate executable that contains utility functions related to the tempo software.  Although it is not required for a working installation, Tempo CLI can be helpful for deeper analysis or for troubleshooting.
```

achatterjee-grafana commented on docs/tempo/website/cli/_index.md:
````diff
@@ -0,0 +1,114 @@
+```
+
+## Running Tempo CLI
+Tempo CLI is currently only available as source and a working go installation is required to build it.  It can be compiled to a native binary and executed normally, or it can be executed using the `go run` command. 
````
```suggestion
Tempo CLI is currently available as a source. A working Go installation is required to build it. It can be compiled to a native binary and executed normally, or it can be executed using the `go run` command. 
```

mdisibio replied:
Updated as discussed

achatterjee-grafana commented on docs/tempo/website/cli/_index.md:
```diff
@@ -0,0 +1,114 @@
+
+## Backend options
+
+Tempo CLI connects directly to the storage backend for some commands, meaning it requires the ability to read from S3, GCS, or file-system storage.  The backend can be configured in a few ways:
```
```suggestion
Tempo CLI connects directly to the storage backend for some commands, meaning that it requires the ability to read from S3, GCS, or file-system storage.  The backend can be configured in a few ways:
```

achatterjee-grafana commented on docs/tempo/website/cli/_index.md:
```diff
@@ -0,0 +1,114 @@
+
+Tempo CLI connects directly to the storage backend for some commands, meaning it requires the ability to read from S3, GCS, or file-system storage.  The backend can be configured in a few ways:
+
+* Load an existing tempo config file using `--config-file` (`-c`) option. This is the recommended option for frequent usage. See [Configuration](../configuration/) documentation for more information.
```
```suggestion
* Load an existing tempo configuration file using the `--config-file` (`-c`) option. This is the recommended option for frequent usage. Refer to [Configuration](../configuration/) documentation for more information.
```

achatterjee-grafana commented on docs/tempo/website/cli/_index.md:
```diff
@@ -0,0 +1,114 @@
+Tempo CLI connects directly to the storage backend for some commands, meaning it requires the ability to read from S3, GCS, or file-system storage.  The backend can be configured in a few ways:
+
+* Load an existing tempo config file using `--config-file` (`-c`) option. This is the recommended option for frequent usage. See [Configuration](../configuration/) documentation for more information.
+* Specify the individual settings:
```
```suggestion
* Specify individual settings:
```

achatterjee-grafana commented on docs/tempo/website/cli/_index.md:
```diff
@@ -0,0 +1,114 @@
+
+* Load an existing tempo config file using `--config-file` (`-c`) option. This is the recommended option for frequent usage. See [Configuration](../configuration/) documentation for more information.
+* Specify the individual settings:
+    * `--backend <value>` The storage backend type, one of `s3`, `gcs`, `local`
```
```suggestion
    * `--backend <value>` The storage backend type, one of `s3`, `gcs`, and `local`.
```

achatterjee-grafana commented on docs/tempo/website/cli/_index.md:
```diff
@@ -0,0 +1,114 @@
+* Load an existing tempo config file using `--config-file` (`-c`) option. This is the recommended option for frequent usage. See [Configuration](../configuration/) documentation for more information.
+* Specify the individual settings:
+    * `--backend <value>` The storage backend type, one of `s3`, `gcs`, `local`
+    * `--bucket <value>` The bucket name.  The meaning of this value is backend specific. See configuration documentation for more information.
```
```suggestion
    * `--bucket <value>` The bucket name. The meaning of this value is backend-specific. Refer to [Configuration](../configuration/) documentation for more information.

achatterjee-grafana commented on docs/tempo/website/cli/_index.md:
```diff
@@ -0,0 +1,114 @@
+    * `--backend <value>` The storage backend type, one of `s3`, `gcs`, `local`
+    * `--bucket <value>` The bucket name.  The meaning of this value is backend specific. See configuration documentation for more information.
+    * `--s3-endpoint <value>` The S3 API endpoint (i.e. s3.dualstack.us-east-2.amazonaws.com).
+    * `--s3-user <value>`, `--s3-password <value>` The S3 user name and password (or access key and secret key). Optional, tempo-cli supports the same authentication mechanisms as tempo. See [S3 permissions documentation](../configuration/s3/#permissions) for more information.
```
```suggestion
    * `--s3-user <value>`, `--s3-password <value>` The S3 user name and password (or access key and secret key). Optional, as Tempo CLI supports the same authentication mechanisms as Tempo. See [S3 permissions documentation](../configuration/s3/#permissions) for more information.
```

achatterjee-grafana commented on docs/tempo/website/cli/_index.md:
```diff
@@ -0,0 +1,114 @@
+- `block-id` The block ID as UUID string.
+
+Options:
+- `--check-dupes` Also load the block data and perform integrity check for duplicates. _Warning:_ can be intense.
```
```suggestion
- `--check-dupes` Also load the block data and perform an integrity check for duplicates. 
_Warning:_ can be intense.
```

mdisibio replied:
As discussed, this was changed to **Note**

achatterjee-grafana commented on docs/tempo/website/cli/_index.md:
```diff
@@ -0,0 +1,114 @@
+- `tenant-id` The tenant ID.  Use `single-tenant` for single tenant setups.
+
+Options:
+- `--load-index` Also load the block indexes and perform integrity checks for duplicates. _Warning:_ can be intense.
```
```suggestion
- `--load-index` Also load the block indexes and perform integrity checks for duplicates. 
_Warning:_ can be intense.
```

achatterjee-grafana commented on docs/tempo/website/cli/_index.md:
```diff
@@ -0,0 +1,114 @@
+
+**Output:**
+Explanation of output:
+- `ID` block ID
+- `Lvl` Compaction level of the block
+- `Count` Number of objects stored in the block
+- `Window` The time window as considered for compaction purposes
+- `Start` The earliest timestamp stored in the block
+- `End` The latest timestamp stored in the block
+- `Duration` Time duration between start and end
+- `Idx` Number of records stored in the index (present when --load-index is specified)
+- `Dupe` Number of duplicate entries in the index. Should be zero.  (present when --load-index is specified)
```
```suggestion
- `ID` Block ID.
- `Lvl` Compaction level of the block.
- `Count` Number of objects stored in the block.
- `Window` The time window considered for compaction purposes.
- `Start` The earliest timestamp stored in the block.
- `End` The latest timestamp stored in the block.
- `Duration` Time duration between start and end.
- `Idx` Number of records stored in the index (present when --load-index is specified).
- `Dupe` Number of duplicate entries in the index. Should be zero.  (present when --load-index is specified).
```

achatterjee-grafana commented on docs/tempo/website/cli/_index.md:
````diff
@@ -0,0 +1,114 @@
+```
+
+Arguments:
+- `api-endpoint` Url for tempo api
````
```suggestion
- `api-endpoint` URL for tempo API.
```

achatterjee-grafana commented on docs/tempo/website/cli/_index.md:
```diff
@@ -0,0 +1,114 @@
+
+Arguments:
+- `api-endpoint` Url for tempo api
+- `trace-id` Trace ID as hexadecimal string
```
```suggestion
- `trace-id` Trace ID as a hexadecimal string.

achatterjee-grafana commented on docs/tempo/website/cli/_index.md:
```diff
@@ -0,0 +1,114 @@
+- `trace-id` Trace ID as hexadecimal string
+
+Options:
+- `--org-id <value>` Organization ID (for use in multi-tenant setup)
```
```suggestion
- `--org-id <value>` Organization ID (for use in multi-tenant setup).
```

achatterjee-grafana reviewed (approved):
Doc content looks good. Added some copy-edit suggestions.
